File: generated_codes/chat_agents/chat-agents-cpp/skiplist/claude_sonnet_4.6/output10_skiplist.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <random>

class ConcurrentDataStructure : public SetADT {
    static constexpr int MAX_LEVEL = 16;

    struct Node {
        int val;
        int topLevel;
        std::atomic<Node*> forward[MAX_LEVEL];
        Node(int v, int lvl) : val(v), topLevel(lvl) {
            for (int i = 0; i < MAX_LEVEL; i++)
                forward[i].store(nullptr, std::memory_order_relaxed);
        }
    };

    static bool is_marked_ref(Node* p) { return (reinterpret_cast<uintptr_t>(p) & 1UL) != 0; }
    static Node* get_unmarked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1UL); }
    static Node* get_marked_ref(Node* p) { return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1UL); }

    Node* head_;
    Node* tail_;
    std::mt19937 rng_;
    std::uniform_int_distribution<int> dist_;

    int randomLevel() {
        int lvl = 1;
        while (lvl < MAX_LEVEL && (dist_(rng_) & 1)) lvl++;
        return lvl;
    }

    void cleanUp(int lv, Node* pred, Node* curr) {
        Node* succ = get_unmarked_ref(curr->forward[lv].load(std::memory_order_acquire));
        Node* expected = curr;
        pred->forward[lv].compare_exchange_strong(expected, succ,
            std::memory_order_acq_rel, std::memory_order_acquire);
    }

    bool find(int key, Node** preds, Node** succs) {
    retry:
        Node* pred = head_;
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            Node* curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
            while (true) {
                Node* succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                while (is_marked_ref(succ_raw)) {
                    cleanUp(lv, pred, curr);
                    Node* raw = pred->forward[lv].load(std::memory_order_acquire);
                    if (is_marked_ref(raw)) goto retry;
                    curr = get_unmarked_ref(raw);
                    succ_raw = curr->forward[lv].load(std::memory_order_acquire);
                }
                if (curr->val < key) {
                    pred = curr;
                    curr = get_unmarked_ref(succ_raw);
                } else break;
            }
            preds[lv] = pred;
            succs[lv] = curr;
        }
        return succs[0]->val == key;
    }

public:
    ConcurrentDataStructure() : rng_(std::random_device{}()), dist_(0, INT_MAX) {
        head_ = new Node(INT_MIN, MAX_LEVEL);
        tail_ = new Node(INT_MAX, MAX_LEVEL);
        for (int i = 0; i < MAX_LEVEL; i++)
            head_->forward[i].store(tail_, std::memory_order_relaxed);
    }

    ~ConcurrentDataStructure() {
        Node* curr = head_;
        while (curr) {
            Node* next = get_unmarked_ref(curr->forward[0].load(std::memory_order_relaxed));
            delete curr;
            curr = next;
        }
    }

    bool contains(int key) override {
        Node* pred = head_;
        Node* curr = nullptr;
        for (int lv = MAX_LEVEL - 1; lv >= 0; lv--) {
            curr = get_unmarked_ref(pred->forward[lv].load(std::memory_order_acquire));
            while (curr->val < key)
                curr = get_unmarked_ref(curr->forward[lv].load(std::memory_order_acquire));
        }
        return curr->val == key && !is_marked_ref(curr->forward[0].load(std::memory_order_acquire));
    }

    bool add(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        int lvl = randomLevel();
        while (true) {
            if (find(key, preds, succs)) return false;
            Node* newNode = new Node(key, lvl);
            for (int i = 0; i < lvl; i++)
                newNode->forward[i].store(succs[i], std::memory_order_relaxed);
            Node* expected = succs[0];
            if (!preds[0]->forward[0].compare_exchange_strong(expected, newNode,
                    std::memory_order_acq_rel, std::memory_order_acquire)) {
                delete newNode; continue;
            }
            for (int i = 1; i < lvl; i++) {
                while (true) {
                    expected = succs[i];
                    if (preds[i]->forward[i].compare_exchange_strong(expected, newNode,
                            std::memory_order_acq_rel, std::memory_order_acquire))
                        break;
                    find(key, preds, succs);
                    newNode->forward[i].store(succs[i], std::memory_order_release);
                }
            }
            return true;
        }
    }

    bool remove(int key) override {
        Node* preds[MAX_LEVEL], *succs[MAX_LEVEL];
        if (!find(key, preds, succs)) return false;
        Node* target = succs[0];
        if (target->val != key) return false;
        for (int i = target->topLevel - 1; i >= 1; i--) {
            Node* succ = target->forward[i].load(std::memory_order_acquire);
            while (!is_marked_ref(succ)) {
                if (target->forward[i].compare_exchange_weak(succ, get_marked_ref(succ),
                        std::memory_order_acq_rel, std::memory_order_acquire))
                    break;
            }
        }
        Node* succ = target->forward[0].load(std::memory_order_acquire);
        while (true) {
            if (is_marked_ref(succ)) return false;
            if (target->forward[0].compare_exchange_strong(succ, get_marked_ref(succ),
                    std::memory_order_acq_rel, std::memory_order_acquire)) {
                find(key, preds, succs);
                return true;
            }
        }
    }
};

File: generated_codes/chat_agents/chat-agents-java/skiplist/gemini_java/output10_skiplist.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicMarkableReference;
import java.util.Random;

public class ConcurrentDataStructure implements SetADT {
    
    private static final int MAX_LEVEL = 16;
    private final Node head;
    private final Node tail;
    private final Random random;

    static final class Node {
        final int key;
        final int height;
        final AtomicMarkableReference<Node>[] next;

        @SuppressWarnings("unchecked")
        Node(int key, int height) {
            this.key = key;
            this.height = height;
            this.next = (AtomicMarkableReference<Node>[]) new AtomicMarkableReference[height + 1];
            for (int i = 0; i <= height; i++) {
                this.next[i] = new AtomicMarkableReference<>(null, false);
            }
        }
    }

    public ConcurrentDataStructure() {
        this.head = new Node(Integer.MIN_VALUE, MAX_LEVEL);
        this.tail = new Node(Integer.MAX_VALUE, MAX_LEVEL);
        for (int i = 0; i <= MAX_LEVEL; i++) {
            head.next[i].set(tail, false);
        }
        this.random = new Random();
    }

    private int randomLevel() {
        int lvl = 0;
        while (lvl < MAX_LEVEL && random.nextBoolean()) {
            lvl++;
        }
        return lvl;
    }

    private boolean find(int key, Node[] preds, Node[] succs) {
        boolean[] marked = {false};
        boolean retry;
        
        retry_loop:
        while (true) {
            Node pred = head;
            for (int level = MAX_LEVEL; level >= 0; level--) {
                Node curr = pred.next[level].getReference();
                while (true) {
                    Node succ = curr.next[level].get(marked);
                    while (marked[0]) {
                        boolean snip = pred.next[level].compareAndSet(curr, succ, false, false);
                        if (!snip) {
                            continue retry_loop;
                        }
                        curr = pred.next[level].getReference();
                        succ = curr.next[level].get(marked);
                    }
                    if (curr.key < key) {
                        pred = curr;
                        curr = succ;
                    } else {
                        break;
                    }
                }
                preds[level] = pred;
                succs[level] = curr;
            }
            return succs[0].key == key;
        }
    }

    @Override
    public boolean add(int key) {
        int topLevel = randomLevel();
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];

        while (true) {
            if (find(key, preds, succs)) {
                return false;
            }
            Node newNode = new Node(key, topLevel);
            for (int i = 0; i <= topLevel; i++) {
                newNode.next[i].set(succs[i], false);
            }
            Node pred = preds[0];
            Node succ = succs[0];
            if (!pred.next[0].compareAndSet(succ, newNode, false, false)) {
                continue;
            }
            for (int i = 1; i <= topLevel; i++) {
                while (true) {
                    pred = preds[i];
                    succ = succs[i];
                    if (pred.next[i].compareAndSet(succ, newNode, false, false)) {
                        break;
                    }
                    find(key, preds, succs);
                }
            }
            return true;
        }
    }

    @Override
    public boolean remove(int key) {
        Node[] preds = new Node[MAX_LEVEL + 1];
        Node[] succs = new Node[MAX_LEVEL + 1];
        
        while (true) {
            if (!find(key, preds, succs)) {
                return false;
            }
            Node nodeToRemove = succs[0];
            for (int level = nodeToRemove.height; level >= 1; level--) {
                boolean[] marked = {false};
                Node succ = nodeToRemove.next[level].get(marked);
                while (!marked[0]) {
                    nodeToRemove.next[level].compareAndSet(succ, succ, false, true);
                    succ = nodeToRemove.next[level].get(marked);
                }
            }
            boolean[] marked = {false};
            Node succ = nodeToRemove.next[0].get(marked);
            while (!marked[0]) {
                if (nodeToRemove.next[0].compareAndSet(succ, succ, false, true)) {
                    // Node has been marked
                    find(key, preds, succs);
                    return true;
                }
                succ = nodeToRemove.next[0].get(marked);
            }
            return false;
        }
    }

    @Override
    public boolean contains(int key) {
        boolean[] marked = {false};
        Node pred = head;
        Node curr = null;
        for (int level = MAX_LEVEL; level >= 0; level--) {
            curr = pred.next[level].getReference();
            while (true) {
                Node succ = curr.next[level].get(marked);
                while (marked[0]) {
                    curr = pred.next[level].getReference();
                    succ = curr.next[level].get(marked);
                }
                if (curr.key < key) {
                    pred = curr;
                    curr = succ;
                } else {
                    break;
                }
            }
        }
        return curr.key == key;
    }
}